bindings/python/DeviceCommissioner.cpp

```diff
@@ -0,0 +1,134 @@
+/*
+ *    Copyright (c) 2021 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include <Python.h>
+
+#include <DeviceCommissioner.h>
+
+#include <controller/CHIPDeviceController.h>
+#include <support/ErrorStr.h>
+
+#include <CommissionerInitParams.h>
+#include <RendezvousParameters.h>
+
+static PyObject * PythonDeviceCommissionerNew(PyTypeObject *type, PyObject *args, PyObject *kwds)
+{
+    uint64_t nodeId;
+    PyObject * initParameters;
+
+    // 1st arg (K: uint64_t): node id
+    // 2nd arg (O!: PyObject *): init parameters
+    if (!PyArg_ParseTuple(args, "KO!", &nodeId, &PythonCommissionerInitParamsType, &initParameters)) {
+        return nullptr;
+    }
+
+    PythonDeviceCommissioner *pyo = reinterpret_cast<PythonDeviceCommissioner*>(type->tp_alloc(type, 0));
+
+    new (&pyo->mCommissioner) chip::Controller::DeviceCommissioner();
+    CHIP_ERROR err = pyo->mCommissioner.Init(nodeId, reinterpret_cast<PythonCommissionerInitParams*>(initParameters)->mPrarms);
+    if (err != CHIP_NO_ERROR) {
+        pyo->~PythonDeviceCommissioner();
+        PyErr_Format(PyExc_RuntimeError, "DeviceCommissioner Init failed: %s.", chip::ErrorStr(err));
+        type->tp_free(pyo);
+        return nullptr;
+    }
+    return (PyObject *)pyo;
+}
+
+static void PythonDeviceCommissionerDealloc(PyObject *self)
+{
+    PythonDeviceCommissioner *pyo = reinterpret_cast<PythonDeviceCommissioner*>(self);
+    pyo->mCommissioner.Shutdown();
+    pyo->mCommissioner.~DeviceCommissioner();
+
+    PyTypeObject *tp = Py_TYPE(self);
+    tp->tp_free(self);
+}
+
+PyDoc_STRVAR(ServiceEventsDocument, "ServiceEvents()");
+static PyObject * PythonDeviceCommissionerServiceEvents(PyObject *self, PyObject *args)
```

Comment:
is ServiceEvents something we would want python code to have access to and use?

Reply:
I'd prefer exactly copy between C++ API and python API. The binding only do type check and type convert, no more other things.

So I'd like to copy all interface function to python.